src/impl/wone/wash.cpp

```diff
@@ -0,0 +1,287 @@
+        assert(!started);
+        init_kernels.push_back(std::make_unique<UpdateKernel>(func));
+    }
+
+    void add_init_void_kernel(const VoidFuncT func) {
+        assert(!started);
+        init_kernels.push_back(std::make_unique<VoidKernel>(func));
+    }
+
+    void add_force_kernel(const ForceFuncT func) {
+        assert(!started);
+        loop_kernels.push_back(std::make_unique<ForceKernel>(func));
+    }
+
+    void add_update_kernel(const UpdateFuncT func) {
+        assert(!started);
+        loop_kernels.push_back(std::make_unique<UpdateKernel>(func));
+    }
+
+    void add_reduction_kernel(const MapFuncT map_func, const ReduceOp reduce_op, double* variable) {
+        assert(!started);
+        loop_kernels.push_back(std::make_unique<ReductionKernel>(map_func, reduce_op, variable));
+    }
+
+    void add_void_kernel(const VoidFuncT func) {
+        assert(!started);
+        loop_kernels.push_back(std::make_unique<VoidKernel>(func));
+    }
+
+    void set_default_neighbor_search(const unsigned max_count) {
+        assert(!started);
+        neighbors_max = max_count;
+        neighbors_kernel = [](Particle& p) { p.recalculate_neighbors(neighbors_max); };
+    }
+
+    void set_neighbor_search_kernel(const NeighborsFuncT func, const unsigned max_count) {
+        assert(!started);
+        neighbors_max = max_count;
+        neighbors_kernel = func;
+    }
+
+    void set_simulation_name(const std::string name) {
+        assert(!started);
+        simulation_name = name;
+    }
+
+    void set_output_file_name(const std::string name) {
+        assert(!started);
+        output_file_name = name;
+    }
+
+    std::vector<Particle>& get_particles() {
+        return local_particles;
+    }
+
+    std::vector<Particle>& get_global_particles() {
+        return particles;
+    }
+
+    std::tuple<int, int> init_mpi() {
+        int rank = 0;
+        int n_ranks = 0;
+        MPI_Init(NULL, NULL);
+        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+        MPI_Comm_size(MPI_COMM_WORLD, &n_ranks);
+
+        return std::make_tuple(rank, n_ranks);
+    }
+
+    cstone::Domain<uint64_t, double, cstone::CpuTag>* init_domain(int rank, int n_ranks, size_t num_particles) {
+        uint64_t bucket_size_focus = 64;
+        // we want about 100 global nodes per rank to decompose the domain with +-1% accuracy
+        uint64_t bucket_size = std::max(bucket_size_focus, num_particles / (100 * n_ranks));
+        float theta = 0.5f;
+        return new cstone::Domain<uint64_t, double, cstone::CpuTag>(
+            rank, n_ranks, bucket_size, bucket_size_focus, theta,
+            cstone::Box(box_xmin, box_xmax, box_ymin, box_ymax, box_zmin, box_zmax, box_xtype, box_ytype, box_ztype));
+    }
+
+    void sync_domain(cstone::Domain<uint64_t, double, cstone::CpuTag>& domain, std::vector<size_t>& keys,
+                     std::vector<double>& s1, std::vector<double>& s2, std::vector<double>& s3) {
+        class _wash_sync_domain;
```

Comment:
These are used by the `ws2st` program to insert code into the backends which requires a reference to e.g. a force array. Mainly used them because an empty class declaration is "inert" and also a really simple matcher to write.